libsolidity/analysis/StaticAnalyzer.h

```diff
@@ -73,8 +73,22 @@ class StaticAnalyzer: private ASTConstVisitor
 	bool visit(BinaryOperation const& _operation) override;
 	bool visit(FunctionCall const& _functionCall) override;
 
+	struct TypeComp
+	{
+		bool operator()(Type const* lhs, Type const* rhs) const
+		{
+			solAssert(lhs && rhs, "");
+			return lhs->richIdentifier() < rhs->richIdentifier();
+		}
+	};
+	using TypeSet = std::set<Type const*, TypeComp>;
```

Comment:
I did write something similar before, and @chriseth mentioned that comparing `Type*` using `richIdentifier()` has a problem that it may not be consistent with `Type::operator==`

Reply:
@hrkrshnn Thanks! I planned  to continue on a dependent PR #9062 during the weekend. So I guess if this one is approved, we can merge it and come back to the `[rich]Identifier` issue when @chriseth is back.

Reply:
Really? Curious. If that's true, we should definitely add a note about it to the declaration in ``Types.h``, which pretty much says the opposite :-). However, in any case ``richIdentifier()`` *will* be consistent with ``identifier()`` and since this here is only used for detailed warning reporting, it should be fine either way.

Reply:
For context: https://github.com/ethereum/solidity/pull/9146#discussion_r442195103

Reply:
I would not want to rely on the identifiers being consistent with `operator<` for correctness, but for constructing the error message, this is fine.

Reply:
Ah yeah, ok - I'd say we have no need for a strong guarantee here, since this is only used for reporting, so I'd say it's fine.